### PR TITLE
Embed dependencies in setup.py

### PR DESCRIPTION
**Pull request by jsiirola**
## Summary/Motivation:
Pyomo currently has dependencies scattered across the GHA / Jenkins builders, `.readthedocs-requirements.txt`, and `setup.py`.  This PR moves all dependency specification into `setup.py`, and updates the GHA and RTD builders to pull dependencies from `setup.py`.  As part of the rework, `setup.py` now specifies three "extras" targets for managing optional dependencies:
- `tests` for dependencies needed to run the Pyomo unit tests (currently, `nose`, `coverage`, `parameterized`, and `pybind11`)
- `docs` for dependencies needed to build the RTD documentation
- `optional` for packages (available on PyPI) on which Pyomo has an optional dependency.

With this, a reasonable approximation of the old `pyomo-extras` package is:
```
pip install 'pyomo[tests,docs,optional]`
```
The challenge with this PR is that `conda` does not play well with `pip` requirements.  The complexity here is to provide a bridge whereby our conda builds can interrogate setup.py to get the list of packages that pip would have otherwise installed (and then conda can install from conda-forge).

## Changes proposed in this PR:
- add `extras_requires` to `setup.py`
- add a new command `python setup.py dependencies` to get lists of required and optional dependency packages
- rework the GHA builders to get dependencies from `setup.py`
- rework RTD configuration to use a `.readthedocs.yaml` configuration and get documentation requirements from `pyomo[docs]`

### Legal Acknowledgement

By contributing to this software project, I have read the [contribution guide](https://pyomo.readthedocs.io/en/stable/contribution_guide.html) and agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the BSD license.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: .github/workflows/test_branches.yml [hunk 1]
```diff
@@ -107,16 +99,12 @@ jobs:
             echo "GHA_JOBGROUP=other" >> $GITHUB_ENV
         fi
         # Note: pandas 1.0.3 causes gams 29.1.0 import to fail in python 3.8
-        PYTHON_PACKAGES="${PYTHON_REQUIRED_PKGS}"
+        EXTRAS=tests
```
Comment (mrmundt) [inline .github/workflows/test_branches.yml:102]: Alternative way to do this: in the `env` section, set `EXTRAS=test,docs,optional` and then for `slim`, change that. I have no strong opinions, but it's fewer lines.

File: .github/workflows/test_branches.yml [hunk 2]
```diff
@@ -276,6 +277,11 @@ jobs:
                     || echo "WARNING: $PKG is not available"
             done
         fi
+        # Re-try Pyomo (optional) dependencies with pip
+        if test -n "$PYPY_DEPENDENCIES"; then
```
Comment (mrmundt) [inline .github/workflows/test_branches.yml:281]: `PYPY` or `PYPI`?

File: .github/workflows/test_pr_and_main.yml [hunk 3]
```diff
@@ -295,6 +296,11 @@ jobs:
                     || echo "WARNING: $PKG is not available"
             done
         fi
+        # Re-try Pyomo (optional) dependencies with pip
+        if test -n "$PYPY_DEPENDENCIES"; then
```
Comment (mrmundt) [inline .github/workflows/test_pr_and_main.yml:300]: `PYPY` or `PYPI`?

File: setup.py [hunk 4]
```diff
@@ -88,22 +92,72 @@ def get_version():
...
-      long_description=read('README.md'),
-      long_description_content_type='text/markdown',
-      keywords=['optimization'],
-      classifiers=[
+
+class DependenciesCommand(Command):
+    """Custom setuptools command
+
+    This will output the list of dependencies, including any optional
+    dependenties for 'extras_require` targets.  This is needed so that
```
Comment (mrmundt) [inline setup.py:100]: Typo: dependencies

File: setup.py [hunk 5]
```diff
@@ -122,31 +176,73 @@ def run_setup():
...
+            'scipy', # Needed by autodoc for pynumero
+        ],
+        'optional': [
+            'dill',      # No direct use, but improves lambda pickle
+            'ipython',   # contrib.viewer
+            'networkx',  # network, incidence_analysis, community_detection
+            'openpyxl',  # dataportals
+            #'pathos',   # requested for #963, but PR currently closed
+            'pint',      # units
+            #'pyro4',    # used by PySP; no longer needed by core Pyomo
```
Comment (mrmundt) [inline setup.py:206]: Is it worth leaving this in here if it's not actually required and there aren't plans for it to be?

File: setup.py [hunk 6]
```diff
@@ -122,31 +176,73 @@ def run_setup():
...
+        ],
+    },
+    packages = find_packages(exclude=("scripts",)),
+    package_data = {
+        "pyomo.contrib.appsi.cmodel": ["src/*"],
+        "pyomo.contrib.mcpp": ["*.cpp"],
+        "pyomo.contrib.pynumero": ['src/*', 'src/tests/*'],
+        "pyomo.contrib.viewer": ["*.ui"],
+    },
+    #include_package_data=True,
```
Comment (mrmundt) [inline setup.py:231]: Do we need this?

**Review by mrmundt — CHANGES_REQUESTED** [review]

File: .github/workflows/test_branches.yml (hunk 2 above)
Comment (jsiirola) [reply .github/workflows/test_branches.yml]: PYPI.  Fixed

File: .github/workflows/test_pr_and_main.yml (hunk 3 above)
Comment (jsiirola) [reply .github/workflows/test_pr_and_main.yml]: PYPI.  Fixed

File: setup.py (hunk 5 above)
Comment (jsiirola) [reply setup.py]: Nope.  Removed.

File: setup.py (hunk 6 above)
Comment (jsiirola) [reply setup.py]: Nope.  Removed.

**Review by mrmundt — APPROVED** [review]

**Pull request merged**